gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/OAuth2SsoConfig.groovy

```diff
@@ -0,0 +1,137 @@
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
+
+/**
+ * Each time a class extends WebSecurityConfigurerAdapter, a new set of matchers + filters gets added
+ * to the FilterChainProxy. This is not great when we want to protect the same URLs with different kinds of
+ * security mechanisms, because only the first set of filters that hit the matchers will be executed.
+ *
+ * In order to get around this, we must only have 1 class that extends WebSecurityConfigurerAdapter.
+ * Unfortunately, the OAuth2 library (sping-security-oauth) has such a class baked in. If we have OAuth2 enabled,
+ * AND want an additional kind of security mechanism (X509, for example), we must use the
+ * {@link OAuth2SsoConfigurerAdapter} hook the library has provided. This way, only a single set of matchers
+ * and filters are added.
+ */
+@Configuration
+@SpinnakerAuthConfig
+// Use @EnableWebSecurity if/when updated to Spring Security 4.
+@EnableWebMvcSecurity
+@Import(SecurityAutoConfiguration)
+@EnableOAuth2Sso
+// Note the 4 single-quotes below - this is a raw groovy string, because SpEL and groovy
+// string syntax overlap!
+@ConditionalOnExpression(''''${spring.oauth2.client.clientId:}'!=""''')
+class OAuth2SsoConfig extends OAuth2SsoConfigurerAdapter {
+
+  @Override
+  void match(OAuth2SsoConfigurer.RequestMatchers matchers) {
+    matchers.antMatchers('/**')
+  }
+
+  @Override
+  void configure(HttpSecurity http) throws Exception {
+    AuthConfig.configure(http)
+  }
+
+  /**
+   * ResourceServerTokenServices is an interface used to manage access tokens. The UserInfoTokenService object is an
+   * implementation of that interface that uses an access token to get the logged in user's data (such as email or
+   * profile). We want to customize the Authentication object that is returned to include our custom (Kork) User.
+   */
+  @Primary
+  @Bean
+  ResourceServerTokenServices spinnakerAuthorityInjectedUserInfoTokenServices() {
+    return new ResourceServerTokenServices() {
+
+      @Autowired
+      ResourceServerProperties sso
+
+      @Autowired
+      UserInfoTokenServices userInfoTokenServices
+
+      @Autowired
+      AccountsService accountsService
+
+      @Autowired
+      UserRolesProvider userRolesProvider
+
+      @Override
+      OAuth2Authentication loadAuthentication(String accessToken) throws AuthenticationException, InvalidTokenException {
+        OAuth2Authentication oAuth2Authentication = userInfoTokenServices.loadAuthentication(accessToken)
+
+        Map details = oAuth2Authentication.userAuthentication.details as Map
+        def email = details.email ?: details.userPrincipalName
+        def roles = userRolesProvider.loadRoles(email)
+
+        // TODO(ttomsu): Pull these values into config so that every OAuth provider doesn't need to specify how to
+        // extract user details.
+        User spinnakerUser = new User(
```

Comment:
What do you have in mind for config? 

Authenticating with GitHub OAuth provider returns me 
```
accountNonExpired: true,
accountNonLocked: true,
allowedAccounts: [
"Teads"
],
authorities: [ ],
credentialsNonExpired: true,
email: "gbougeard@gmail.com",
enabled: true,
firstName: null,
lastName: null,
roles: [ ],
username: "gbougeard@gmail.com"
}
```
But for authorization part, I would need the GitHub login (as `username` I think) to be able to retrieve informations about the user.